src/python/pants/engine/scheduler.py

```diff
@@ -164,37 +161,36 @@ def _to_ids_buf(self, types):
     def _to_utf8_buf(self, string):
         return self._native.context.utf8_buf(string)
 
-    def _register_rules(self, rule_index):
-        """Record the given RuleIndex on `self._tasks`."""
+    def _register_rules(self, rule_index: RuleIndex) -> Any:
+        """Create a native Tasks object, and record the given RuleIndex on it."""
+
+        tasks = self._native.new_tasks()
         registered = set()
+
         for output_type, rules in rule_index.rules.items():
             for rule in rules:
                 key = (output_type, rule)
-                if key in registered:
-                    continue
                 registered.add(key)
 
                 if type(rule) is TaskRule:
-                    self._register_task(output_type, rule, rule_index.union_rules)
+                    self._register_task(tasks, output_type, rule, rule_index.union_rules)
                 else:
                     raise ValueError("Unexpected Rule type: {}".format(rule))
 
-    def _register_task(self, output_type, rule: TaskRule, union_rules):
+        return tasks
+
+    def _register_task(self, tasks: Any, output_type, rule: TaskRule, union_rules) -> None:
```

Comment:
Ditto on leaving off the type for `tasks`. Also, do we know `union_rules` based off of your prior PR?